### DISCUSSION TITLE
Is SWAFAN Supported?

### DISCUSSION THREAD

**Discussion by drewendly** [devcompl/TTController#211] [Configuration]
Are the new SWAFAN 12 and 14 supported? Not getting any controller detected. Only HID and Sensors.

**Comment by szemkhel** [comment]
Same question, not sure if my configuration is faulty or its not supported.
I haven't seen it in support list per se - but Swafan seem to use same controller as the other that appears on supported list.

**Reply by szemkhel** [reply]
there is a picture of controller on this site https://www.thermaltake.com/swafan-12-rgb-radiator-fan-3-fan-pack.html

**Comment by szemkhel** [comment]
@MoshiMoshi0 I couldn't add branch to your repo I guess branching is only for insiders. Anyway, I have forked it [here](https://github.com/szemkhel/TTController) 
I managed to add `Swafan controller`, and with some basic config my `Swafans` are changing colors 🥳 - Not all rings just yet, but maybe its enough for you to figure out support for `Swafan`? I have hardcoded `ProduictId` enumerator, seems to be easy to guess other potential `Ids` If someone will use more than three I have.
Please bare in mind that this is my first encounter with CS language, developing with .NET, or developing on windows - I'm a Linux guy.

**Reply by MoshiMoshi0** [reply]
You cant create branches on someones repo, you have to do a PR. Create a separate branch on your repo and commit to that branch, then do a PR. I would not recommend doing PR straight from master.

> Not all rings just yet, but maybe its enough for you to figure out support for Swafan

Figuring out a new controller requires either disassembling the official thermaltake software (what I've always done), or capturing usb data with wireshark + usbpcap.
Since you have the controller you could try capturing the raw data and figuring out the commands.

Also you copied code for Trio controller, but I think the SetRgb command should be more similar to Quad controller. Maybe thats why not all rings are lighting up.

> I have hardcoded ProduictId enumerator, seems to be easy to guess other potential Ids If someone will use more than three I have.

I would just assume that there are 16 pids like for other controllers.

> Please bare in mind that this is my first encounter with CS language, developing with .NET, or developing on windows - I'm a Linux guy.

LGTM for a POC. 
It should be possible to update this project to .net7 and then it would be fully cross platform. You can write C# on linux nowadays without any problems.

**Reply by szemkhel** [reply]
> You cant create branches on someones repo, you have to do a PR. Create a separate branch on your repo and commit to that branch, then do a PR. I would not recommend doing PR straight from master.

That was my assumption, but I'm getting some permission error while trying to push branch.
![Screenshot 2023-08-17 000220](https://github.com/MoshiMoshi0/TTController/assets/8437578/5e326870-a823-4a88-9527-f5976353cb79)

> Figuring out a new controller requires either disassembling the official thermaltake software (what I've always done), or capturing usb data with wireshark + usbpcap.
> Since you have the controller you could try capturing the raw data and figuring out the commands.

I'll give it a try. Thanks for guidance.

**Comment by szemkhel** [comment]
I managed to light all LED's but with this rather ambiguous config:
```
{
  "Profiles": [{
  "Name": "AllFans",
  "Ports": [
    [9802, 9005, 1],
    [9802, 9005, 2],
    [9802, 9005, 3],
    [9802, 9003, 1],
    [9802, 9003, 2],
    [9802, 9003, 3],
    [9802, 9004, 1],
    [9802, 9004, 2],
    [9802, 9004, 3]
  ],
  "Config": {
    "DeviceType": "SwafanTrio",
    "IgnoreColorCache": true
  },
  "Effects": [
    {
      "Type": "StaticColorEffect",
      "Config": {
        "Color": {
          "PerLed": [
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255],
          [255, 255, 255]
          ]
        },
        "ColorGenerationMethod": "PerPort",
        "Trigger": {
          "Type": "AlwaysTrigger"
        }
      }
    }
  ]
  }]
}
```

I tried zone config but it lits only main ring:
```
{
  "Profiles": [{
  "Name": "AllFans",
  "Ports": [
    [9802, 9005, 1],
    [9802, 9005, 2],
    [9802, 9005, 3],
    [9802, 9003, 1],
    [9802, 9003, 2],
    [9802, 9003, 3],
    [9802, 9004, 1],
    [9802, 9004, 2],
    [9802, 9004, 3]
  ],
  "Config": {
    "DeviceType": "SwafanTrio",
    "IgnoreColorCache": true
  },
   "Effects": [
    {
      "Type": "ZoneEffect",
      "Config": {
        "Effects": [
          {
            "Type": "StaticColorEffect",
            "Config": {
              "Color": {
                "Full": [255, 0, 0]
              }
            }
          },
          {
            "Type": "StaticColorEffect",
            "Config": {
              "Color": {
                "Full": [0, 255, 0]
              }
            }
          },
          {
            "Type": "StaticColorEffect",
            "Config": {
              "Color": {
                "Full": [0, 0, 255]
              }
            }
          }
        ],
        "Trigger": {
          "Type": "AlwaysTrigger"
        }
      }
    }
  ]

}]
}
```

I could use some guidance on where to bite it. Again - it's my second day with C# so I'm green as G in RGB. 
Appreciate you patience, TIA
